torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+  TORCH_INTERNAL_ASSERT(vec_expr->outputs().front()->isA<kir::TensorView>());
+  TORCH_INTERNAL_ASSERT(vec_expr->inputs().front()->isA<kir::TensorView>());
+
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+
+  // It is assumed that either of input and output is on global memory
+  TORCH_INTERNAL_ASSERT(
+      out_tv->memoryType() == MemoryType::Global ||
+          in_tv->memoryType() == MemoryType::Global,
+      "Either input or output tensor must be on global memory.");
+  // However, not both of them
+  TORCH_INTERNAL_ASSERT(
+      !(out_tv->memoryType() == MemoryType::Global &&
+        in_tv->memoryType() == MemoryType::Global),
+      "Both input and output tensors are on global memory.");
+  // Must be either global or local
+  TORCH_INTERNAL_ASSERT(
+      (out_tv->memoryType() == MemoryType::Global ||
+       out_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of output tensor");
+  TORCH_INTERNAL_ASSERT(
+      (in_tv->memoryType() == MemoryType::Global ||
+       in_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of input tensor");
+
+  // TensorView on global memory. This is the tensor that may have
+  // a non-aligned base address.
+  auto global_tv =
+      (out_tv->memoryType() == MemoryType::Global) ? out_tv : in_tv;
+
+  // TensorView with the misaligned vec iterDomain. It is the consumer
+  // of vectorized load or the producer of vectorized store. It is
+  // assumed that when the output TV is not on global memory, this
+  // expression is a vectorized load, so the output TV is vec_tv.
+  // TODO: Check vec_tv has indeed MisalignedVectorize parallel type.
+  auto vec_tv = (out_tv->memoryType() != MemoryType::Global) ? out_tv : in_tv;
+
+  auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+  if (pred == nullptr) {
+    pred = ir_builder.create<kir::Bool>(true);
+  }
+
+  kir::IfThenElse* pred_ite = ir_builder.create<kir::IfThenElse>(pred);
+  new_loop->body().push_back(pred_ite);
+
+  // Generate vectorize index
+  auto indices = (out_tv->memoryType() == MemoryType::Global)
+      ? Index::getConsumerStridedIndices(out_tv->fuserTv(), loop_structure)
+      : Index::getProducerStridedIndices(
+            in_tv->fuserTv(), out_tv->fuserTv(), loop_structure);
+
+  //  Get full extent for merged root domains
+  auto extent = getVectorizeExtent(vec_tv, indices);
+
+  auto vector_size =
+      vec_tv->domain()->domain().back()->extent()->as<kir::Int>();
+
+  auto index =
+      ir_builder.create<kir::TensorIndex>(global_tv->fuserTv(), indices);
```

Comment:
We should not be doing the direct indexing in an unrolling pass. This is taking a step in the wrong direction. Quite honestly we should be getting predicate computation out of unrolling.

Reply:
This is the approach you proposed in February instead of using only for-loops.

From 2/8/2021:
> it has a fixed extent on the for loop
otherwise good luck inferring your thread bindings